### PR TITLE
added mode

### PR DESCRIPTION
**Pull request by balast**
- [x] Tests added / passed ~~(Added a few which passes, one fails during pytest, but runs fine otherwise, and I'm stumped)~~
- [x] Passes `black dask` / `flake8 dask`

Adds the mode function for Series and DataFrame, doesn't support numeric_only parameter currently, though I believe I could add it.  Hoping to get some feedback generally and specifically on why my last test won't pass when run in pytest but runs fine otherwise. It's commented out at the moment.

### CODE REVIEW

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -1494,6 +1495,12 @@ def _reduction_agg(self, name, axis=None, skipna=True, split_every=False, out=No
                 result.divisions = (self.columns.min(), self.columns.max())
             return handle_out(out, result)
 
+    def _mode_chunk(df, dropna):
+        if dropna:
+            return Counter(df.dropna().values)
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: Can you instead use `df.value_counts(dropna=dropna)`?

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -1654,6 +1661,44 @@ def count(self, axis=None, split_every=False):
                 result.divisions = (self.columns.min(), self.columns.max())
             return result
 
+    @derived_from(pd.DataFrame)
+    def mode(self, axis=0, dropna=True, split_every=False):  # this is the series class
+        """Use a counter on each partition, sum the resulting counters, iterate over key,
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: I don't think we need this docstring. We'll copy it from DataFrame.mode.

File: dask/dataframe/core.py [hunk 3]
```diff
@@ -1654,6 +1661,44 @@ def count(self, axis=None, split_every=False):
...
+        value pairs until num_occurrences isn't equal to the first one"""
+        if axis == 1:
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: Users can also pass `axis='columns'` You should normalize the axis with `axis = self._validate_axis(axis)`.

File: dask/dataframe/core.py [hunk 4]
```diff
@@ -1654,6 +1661,44 @@ def count(self, axis=None, split_every=False):
...
+                token=self._token_prefix + "mode",
+                axis=axis,
+                dropna=True,
+                enforce_metadata=False,
+            )
+            return result
+        else:
+            from collections import Counter
+
+            def _mode_chunk(df, dropna=dropna):  # already defined elsewhere now
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: Ideally would remove this to avoid the closure.

File: dask/dataframe/core.py [hunk 5]
```diff
@@ -1654,6 +1661,44 @@ def count(self, axis=None, split_every=False):
...
+                aggregate=M.sum,
+                meta="object",
+                split_every=split_every,
+            ).to_delayed()
+            # now I have the all_values_dict as a delayed object.  Next, find max value and then all keys with max value
+            max_val = delayed(lambda x: max(x.values()))(all_values_dict)
+            mode_list = delayed(
+                lambda x, max_val: sorted([k for k, v in x.items() if v == max_val])
+            )(all_values_dict, max_val)
+            return mode_list
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: What's the return type here? This should return a DataFrame I think.

File: dask/dataframe/core.py [hunk 6]
```diff
@@ -4132,6 +4191,45 @@ def applymap(self, func, meta="__no_default__"):
     def round(self, decimals=0):
         return elemwise(M.round, self, decimals)
 
+    @derived_from(pd.DataFrame)
+    def mode(self, axis=0, numeric_only=False, dropna=True, split_every=False):
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]: Hmm why are you defining this on all of _Frame, Series, and DataFrame? I don't think you'd need all three.

File: dask/dataframe/tests/test_dataframe.py [hunk 7]
```diff
@@ -3615,6 +3615,38 @@ def test_dataframe_reductions_arithmetic(reduction):
     )
 
 
+@pytest.mark.filterwarnings("ignore:UserWarning")
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_dataframe.py]: Why are we ignoring warnings here? I don't think we want to.

File: dask/dataframe/tests/test_dataframe.py [hunk 8]
```diff
@@ -3615,6 +3615,38 @@ def test_dataframe_reductions_arithmetic(reduction):
...
+    # test series with object dtype
+    assert_eq(ddf.Name.mode(), df.Name.mode())
+
+    # test series with numeric dtype
+    assert_eq(ddf.Name.mode(), df.Name.mode())
+
+    # test dataframe axis=0
+    assert_eq(ddf.mode(), df.mode())
+
+    # # test dataframe axis=1
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_dataframe.py]: Will need to uncomment these.

File: dask/dataframe/core.py (hunk 1 above)
Comment (balast) [reply dask/dataframe/core.py]: Thanks for suggesting that.  I wasn't aware of value_counts, but looks like it will simplify things quite a bit.

**Comment by balast** [comment]
Closing in favor of https://github.com/dask/dask/pull/5983

File: dask/dataframe/core.py [hunk 9]
```diff
@@ -4132,6 +4171,33 @@ def applymap(self, func, meta="__no_default__"):
...
+
+            mode_series_list = []
+            for col in self.columns:
+                mode_series_list.append(
+                    super(Series, self[col]).mode(
+                        dropna=dropna, split_every=split_every
+                    )
+                )
+
+            ddf = concat(mode_series_list, axis=1)
```
Comment (balast) [inline dask/dataframe/core.py]: @TomAugspurger Thank you for your response.  I wasn't aware of value_counts, but it promises to simplify things quite a bit.  I'm currently implementing mode using value_counts, but I'm hitting an error here on `ddf = concat(mode_series_list, axis=1)`.  The error is below.
```python3  
File "/home/balast/PycharmProjects/dask/dask_repo/dask/dataframe/multi.py", line 865, in concat_and_check
    raise ValueError("Concatenated DataFrames of different lengths")
ValueError: Concatenated DataFrames of different lengths
```
There is no error when concatenating series of different lengths if the divisions are known, but the divisions aren't known here. I believe I can get around this using the delayed api similar to what I did in the last PR, where I add np.nans to each short array until they are all the same length, but before I do so, I was wondering if you have a suggestion to concatenate series of various lengths without using the delayed api.  Things will be simpler if it's possible.

Comment (TomAugspurger) [reply dask/dataframe/core.py]: I don't think you should be concatenating on axis=1. The idea is to define mode in terms of a value_counts over the entire dataset followed by a selection of the index values where the count equals the max. Does that make sense?

So I don't think the lines https://github.com/dask/dask/pull/5958/files#diff-492da7893fa5fa6ad3a6ef6cdef985f2R4184-R4190 are wanted.

Comment (balast) [reply dask/dataframe/core.py]: > The idea is to define mode in terms of a value_counts over the entire dataset followed by a selection of the index values where the count equals the max.

To be consistent with pandas, the mode is calculated along each row or column of the dataframe (depending on the value of axis).  So in the current implementation _Frame.mode does what you describe for a single series.  Then DataFrame.mode() merely calls _Frame.mode for each series in itself (when axis=1).  

The mode of each series can be a different number of values so each series is of different length.  I'm getting an error when trying to concatenate the various length series together into a single dataframe.

Example of Pandas Functionality
-------------------------------------------
```python3 
import pandas as pd 
df = pd.DataFrame({'a': [1,1,2,3], 'b': [2,2,3,3], 'c': [3,4,5,6]})
mode_result = df.mode(axis=0) 
```
image

Comment (TomAugspurger) [reply dask/dataframe/core.py]: Gotcha. I'm not sure what's best then.

Comment (balast) [reply dask/dataframe/core.py]: No worries, I think I can manage with the delayed api

Comment (balast) [reply dask/dataframe/core.py]: I made another PR https://github.com/dask/dask/pull/5990 that will make this possible without resorting to the delayed API.  All tests pass on that PR.

File: dask/dataframe/core.py (hunk 1 above)
Comment (balast) [reply dask/dataframe/core.py]: Now things are using value_counts

File: dask/dataframe/core.py (hunk 2 above)
Comment (balast) [reply dask/dataframe/core.py]: docstring removed

File: dask/dataframe/core.py (hunk 4 above)
Comment (balast) [reply dask/dataframe/core.py]: removed

File: dask/dataframe/core.py (hunk 5 above)
Comment (balast) [reply dask/dataframe/core.py]: No longer applicable

File: dask/dataframe/tests/test_dataframe.py (hunk 8 above)
Comment (balast) [reply dask/dataframe/tests/test_dataframe.py]: No longer applicable

File: dask/dataframe/tests/test_dataframe.py (hunk 7 above)
Comment (balast) [reply dask/dataframe/tests/test_dataframe.py]: Gone

File: dask/dataframe/core.py (hunk 9 above)
Comment (balast) [reply dask/dataframe/core.py]: The other PR proved to be more complex than anticipated.  I closed it, and instead added to this PR to produce the concatenation appropriately for mode.

File: dask/dataframe/core.py (hunk 3 above)
Comment (balast) [reply dask/dataframe/core.py]: The axis are validated in the DataFrame mode method which calls this method so it wouldn't be necessary to _validate_axis again.

File: dask/dataframe/core.py (hunk 6 above)
Comment (balast) [reply dask/dataframe/core.py]: Now mode is nontrivially defined for _Frame and DataFrame.  The Series mode merely calls _Frame mode.

File: dask/dataframe/tests/test_dataframe.py [hunk 10]
```diff
@@ -3615,6 +3615,38 @@ def test_dataframe_reductions_arithmetic(reduction):
...
+def test_dataframe_mode():
+    data = [["Tom", 10, 7], ["Farahn", 14, 7], ["Julie", 14, 5], ["Nick", 10, 10]]
+
+    df = pd.DataFrame(data, columns=["Name", "Age", "Num"])
+    ddf = dd.from_pandas(df, npartitions=3)
+
+    # test series with object dtype
+    assert_eq(
+        ddf.Name.mode().compute().sort_values().reset_index(drop=True), (df.Name.mode())
+    )
```
Comment (balast) [inline dask/dataframe/tests/test_dataframe.py]: This mode implementation doesn't sort the returned values as pandas does so when I compare the results I sort_values and reset_index first.

File: dask/dataframe/core.py [hunk 11]
```diff
@@ -1654,6 +1654,37 @@ def count(self, axis=None, split_every=False):
...
+                chunk_kwargs={"dropna": dropna},
+            )
+
+            max_val = value_count_series.max(skipna=dropna)
+            mode_series = (
+                value_count_series[value_count_series == max_val]
+                .index.to_series()
+                .reset_index(drop=True)
+            )
+            return mode_series
```
Comment (balast) [inline dask/dataframe/core.py]: I currently don't sort the mode values as pandas does, but thinking more about it, I don't believe sorting would require any additional shuffling since by the time the sorting would be done, all the data is in a single partition (if I understand the reduction step correctly).  I'd be happy to add sorting unless there is a reason not too. @TomAugspurger, any thoughts?

**Comment by balast** [comment]
@TomAugspurger this should be ready for another review whenever you get a chance.  Thank you!

**Comment by jrbourbeau** [comment]
@balast just an FYI the Windows CI failures are unrelated to the changes here. I'll work on fixing those CI builds

**Comment by balast** [comment]
@TomAugspurger All checks pass now!  This should be ready for review now!

File: dask/dataframe/tests/test_dataframe.py (hunk 10 above)
Comment (mrocklin) [reply dask/dataframe/tests/test_dataframe.py]: Ideally we would have the following:

```python
assert_eq(ddf.Name.mode(), df.Name.mode())
assert_eq(ddf.mode(), df.mode())
```

Yes, I agree with your recent commend that it should be fine to call operations like sorting and reset index on the single partition result value.

It is important to not call `.compute` in `assert_eq`.  This lets us run a variety of important tests on the Dask object.

File: dask/dataframe/core.py [hunk 12]
```diff
@@ -4132,6 +4181,38 @@ def applymap(self, func, meta="__no_default__"):
...
+            for col in self.columns:
+                mode_series = super(Series, self[col]).mode(
+                    dropna=dropna, split_every=split_every
+                )
+                mode_series.name = col
+                mode_series_list.append(mode_series)
+
+            name = "concat-" + tokenize(*mode_series_list)
+
+            concat_axis1 = partial(methods.concat, axis=1)
```
Comment (mrocklin) [inline dask/dataframe/core.py]: We try to avoid placing partials and other dynamicly generated functions into the task graph.  This aids future debugging and reduces scheduling overhead.  If you need to add a keyword, then consider using the `apply` function.  

However, I also wonder if there is a way to do what you're trying to do here with `map_partitions` instead.  If 
so, it would be good to do so.  More future devs/reviewers will be able to understand `map_partitions` than custom graphs (which are used only when we need to do something complicated).

File: dask/dataframe/core.py (hunk 11 above)
Comment (mrocklin) [reply dask/dataframe/core.py]: I recommend that you create a function that does all of the final assembly bits and then return something like the following:

```python
return self.reduction(
    chunk=M.value_counts,
    combine=M.sum,
    aggregate=_mode_aggregate,  # (you'll make this function somewhere in this module, see _count_aggregate for an example
    ...
)
```

Comment (balast) [reply dask/dataframe/core.py]: I added the _mode_aggregate function as recommended.

File: dask/dataframe/tests/test_dataframe.py (hunk 10 above)
Comment (balast) [reply dask/dataframe/tests/test_dataframe.py]: @mrocklin I've removed the .compute() in testing locally and found an issue which I hadn't noticed before.

The issue is when computing mode for each row of a dataframe (`df.mode(axis=1)`), the number of columns returned isn't known until after mode is computed. (see image)
image

This prevents me from specifying metadata in the typical way when axis=1.  Is there a way to handle a variable number of columns in dask dataframe metadata?  If not, I think I'll leave the axis parameter unimplemented in the mode method for dask's dataframes.  Does that seem reasonable?

Comment (balast) [reply dask/dataframe/tests/test_dataframe.py]: I removed the `axis` parameter from mode and updated tests to not call compute.

File: dask/dataframe/core.py (hunk 12 above)
Comment (balast) [reply dask/dataframe/core.py]: I replaced partial with apply.  

I don't think map_partitions is appropriate here though I'm happy to try additional suggestions.

**Why I don't believe map_partitions is appropriate here**
In this implementation, mode is computed on each series of the dataframe.  As part of computing mode on a series, all of the Series partitions are collapsed into a single partition via `value_counts` and `sum`.  This last bit of the mode method is trying to concatenate the result of having run mode on each of the series.  Each of the series are often of differing lengths, and I am trying to concatenate them along axis=1 to construct the final dataframe to be returned with nans filling in the extra space in the shorter series (as pandas does).

The `concat_unindexed_dataframes` in `dask/dataframe/multi.py` (see below) is very close to what I want to do except it calls `concat_and_check` which checks that the series are the same length, which they are not.  I therefore copied `concat_and_check` and modified it slightly to do what I want in this bit of code.

**Part of `dask/dataframe/multi.py` for reference**
```python3
def concat_and_check(dfs):
    if len(set(map(len, dfs))) != 1:
        raise ValueError("Concatenated DataFrames of different lengths")
    return methods.concat(dfs, axis=1)


def concat_unindexed_dataframes(dfs):
    name = "concat-" + tokenize(*dfs)

    dsk = {
        (name, i): (concat_and_check, [(df._name, i) for df in dfs])
        for i in range(dfs[0].npartitions)
    }

    meta = methods.concat([df._meta for df in dfs], axis=1)

    graph = HighLevelGraph.from_collections(name, dsk, dependencies=dfs)
    return new_dd_object(graph, name, meta, dfs[0].divisions)
```

File: dask/dataframe/core.py [hunk 13]
```diff
@@ -4132,6 +4165,38 @@ def applymap(self, func, meta="__no_default__"):
...
+    def mode(self, axis=0, numeric_only=False, dropna=True, split_every=False):
+        axis = self._validate_axis(axis)
+        if axis == 1:
+            return super(DataFrame, self).mode(
+                axis=axis, dropna=dropna, split_every=split_every
+            )
+        elif axis == 0:
+            mode_series_list = []
+            for col in self.columns:
+                mode_series = super(Series, self[col]).mode(
```
[the remark this answers is not in the document]
Comment (balast) [reply dask/dataframe/core.py]: In this implementation, the mode logic for a Series is implemented within the `_Frame.mode` method.  `DataFrame.mode` merely calls the `_Frame.mode` method and concatenates the resulting series together to the final dataframe.  That is what is occurring in this bit of code and why I am using super here.

File: dask/dataframe/core.py [hunk 14]
```diff
@@ -4132,6 +4165,38 @@ def applymap(self, func, meta="__no_default__"):
     def round(self, decimals=0):
         return elemwise(M.round, self, decimals)
 
+    @derived_from(pd.DataFrame)
+    def mode(self, axis=0, numeric_only=False, dropna=True, split_every=False):
+        axis = self._validate_axis(axis)
+        if axis == 1:
+            return super(DataFrame, self).mode(
```
[the remark this answers is not in the document]
Comment (balast) [reply dask/dataframe/core.py]: This bit is outdated now.

File: dask/dataframe/core.py [hunk 15]
```diff
@@ -2782,6 +2811,10 @@ def groupby(self, by=None, **kwargs):
     def count(self, split_every=False):
         return super(Series, self).count(split_every=split_every)
 
+    @derived_from(pd.Series)
+    def mode(self, dropna=True, split_every=False):
+        return super(Series, self).mode(dropna=dropna, split_every=split_every)
```
[the remark this answers is not in the document]
Comment (balast) [reply dask/dataframe/core.py]: I believe everything will still work correctly if I make this change, but I was trying to maintain consistency with other methods in `dataframe/core.py`  E.g. count (see below)
```python3
@derived_from(pd.Series)
    def count(self, split_every=False):
        return super(Series, self).count(split_every=split_every)
```

**Comment by balast** [comment]
@TomAugspurger I would welcome additional feedback on the latest changes when you are able.  Thank you!

File: dask/dataframe/core.py [hunk 16]
```diff
@@ -4132,6 +4148,35 @@ def applymap(self, func, meta="__no_default__"):
     def round(self, decimals=0):
         return elemwise(M.round, self, decimals)
 
+    @derived_from(pd.DataFrame)
+    def mode(self, dropna=True, split_every=False):
+        mode_series_list = []
+        for col in self.columns:
+            mode_series = super(Series, self[col]).mode(
```
Comment (TomAugspurger) [inline dask/dataframe/core.py]:
```suggestion
            mode_series = self[col].mode(
```

Seems more future proof to call Series.mode, in case it's implementation changes in the future.

Comment (TomAugspurger) [reply dask/dataframe/core.py]: This also may run into issues with duplicate column names, since the `self[col]` would return a DataFrame rather than a series. Perhaps safest to do positional slices or iterate over the columns.

File: dask/dataframe/tests/test_dataframe.py [hunk 17]
```diff
@@ -3615,6 +3615,16 @@ def test_dataframe_reductions_arithmetic(reduction):
...
 
+def test_dataframe_mode():
+    data = [["Tom", 10, 7], ["Farahn", 14, 7], ["Julie", 14, 5], ["Nick", 10, 10]]
+
+    df = pd.DataFrame(data, columns=["Name", "Age", "Num"])
+    ddf = dd.from_pandas(df, npartitions=3)
+
+    assert_eq(ddf.mode(), df.mode())
+    assert_eq(ddf.Name.mode(), df.Name.mode())
+
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_dataframe.py]: Can you add a test for an empty dataframe? `df = pd.DataFrame(columns=['a', 'b'])`.

Comment (TomAugspurger) [reply dask/dataframe/tests/test_dataframe.py:3626]: What's the failure with `check_index=True`? I also suspect it's not a problem but want to verify.

Comment (balast) [reply dask/dataframe/tests/test_dataframe.py:3626]: Yeah, no worries, the error as reported by pytest is shown below.

```python3
E           AssertionError: DataFrame.index are different
E           
E           DataFrame.index classes are not equivalent
E           [left]:  Int64Index([], dtype='int64')
E           [right]: Index([], dtype='object')
```

Comment (TomAugspurger) [reply dask/dataframe/tests/test_dataframe.py:3626]: Thanks. I opened https://github.com/pandas-dev/pandas/issues/33321 in pandas. Can you add a reference to that as a comment so we can remove the `check_index=False` in the future?

Comment (balast) [reply dask/dataframe/tests/test_dataframe.py:3626]: added

**Comment by balast** [comment]
The failing test is test_cov.  I believe I'm seeing https://github.com/dask/dask/issues/5910.  Recommitting to get around this.

**Comment by TomAugspurger** [comment]
Thanks @balast!
